test/integration/controlplane/kms_transformation_test.go

```diff
@@ -176,14 +178,101 @@ resources:
 		t.Fatalf("expected %q after decryption, but got %q", secretVal, string(plainSecret))
 	}
 
+	secretClient := test.restClient.CoreV1().Secrets(testNamespace)
 	// Secrets should be un-enveloped on direct reads from Kube API Server.
-	s, err := test.restClient.CoreV1().Secrets(testNamespace).Get(context.TODO(), testSecret, metav1.GetOptions{})
+	s, err := secretClient.Get(context.TODO(), testSecret, metav1.GetOptions{})
 	if err != nil {
 		t.Fatalf("failed to get Secret from %s, err: %v", testNamespace, err)
 	}
 	if secretVal != string(s.Data[secretKey]) {
 		t.Fatalf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
 	}
+
+	// write data using AES GCM to simulate a downgrade
+	futureSecretBytes, err := base64.StdEncoding.DecodeString(futureSecret)
+	if err != nil {
+		t.Fatalf("failed to base64 decode future secret, err: %v", err)
+	}
+	futureKeyBytes, err := base64.StdEncoding.DecodeString(futureAESGCMKey)
+	if err != nil {
+		t.Fatalf("failed to base64 decode future key, err: %v", err)
+	}
+	block, err := aes.NewCipher(futureKeyBytes)
+	if err != nil {
+		t.Fatalf("invalid key, err: %v", err)
+	}
+
+	// we cannot precompute this because the authenticated data changes per run
+	futureEncryptedSecretBytes, err := aestransformer.NewGCMTransformer(block).TransformToStorage(context.TODO(), futureSecretBytes, value.DefaultContext(secretETCDPath))
+	if err != nil {
+		t.Fatalf("failed to encrypt future secret, err: %v", err)
+	}
+
+	futureEncryptedSecretBuf := cryptobyte.NewBuilder(nil)
+	futureEncryptedSecretBuf.AddBytes([]byte(wantPrefix))
+	futureEncryptedSecretBuf.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes([]byte(futureAESGCMKey))
+	})
+	futureEncryptedSecretBuf.AddBytes(futureEncryptedSecretBytes)
+
+	_, err = test.writeRawRecordToETCD(secretETCDPath, futureEncryptedSecretBuf.BytesOrPanic())
+	if err != nil {
+		t.Fatalf("failed to write future encrypted secret, err: %v", err)
+	}
+
+	// confirm that direct AES CBC decryption does not work
+	failingRawEnvelope, err := test.getRawSecretFromETCD()
+	if err != nil {
+		t.Fatalf("failed to read %s from etcd: %v", secretETCDPath, err)
+	}
+	failingFutureEnvelope := envelope{
+		providerName: providerName,
+		rawEnvelope:  failingRawEnvelope,
+		plainTextDEK: futureKeyBytes,
+	}
+	failingFuturePlainSecret, err := failingFutureEnvelope.plainTextPayload(secretETCDPath)
+	if err == nil || err.Error() != "failed to transform from storage via AESCBC, err: the stored data is not a multiple of the block size" {
```

Comment:
let's reference https://github.com/kubernetes/kubernetes/blob/3e60727f756201f63f22839523c73d7648f47948/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go#L99 instead in case the raw content changes in the future.

Reply:
good idea! Updated.